pkg/kapis/servicemesh/metrics/v1alpha2/handler.go

```diff
@@ -17,122 +17,182 @@ limitations under the License.
 package v1alpha2
 
 import (
+	"context"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"net/http"
 
 	"github.com/emicklei/go-restful"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/klog"
 
 	"kubesphere.io/kubesphere/pkg/api"
+	"kubesphere.io/kubesphere/pkg/simple/client/cache"
+	"kubesphere.io/kubesphere/pkg/simple/client/kiali"
+	"kubesphere.io/kubesphere/pkg/simple/client/servicemesh"
 )
 
-// default jaeger query api endpoint address
-var JaegerQueryUrl = "http://jaeger-query.istio-system.svc:16686"
-
-/*
-Use Kiali API directly if config existed in configmap.
-Such as:
-kubectl -n kubesphere-system get cm kubesphere-config -oyaml
-...
-kialiQueryHost: http://kiali.istio-system:20001
-...
-
-Otherwise, use the API provided by kiali code.
+const (
+	KubesphereNamespace      = "kubesphere-system"
+	KubeSphereServiceAccount = "kubesphere"
+)
 
-Announce: The API provided by kiali code will deprecated in the future.
-*/
+type Handler struct {
+	opt    *servicemesh.Options
+	client *kiali.Client
+}
 
-var KialiQueryUrl string
+func NewHandler(o *servicemesh.Options, client kubernetes.Interface, cache cache.Interface) *Handler {
+	if o != nil && o.KialiQueryHost != "" {
+		sa, err := client.CoreV1().ServiceAccounts(KubesphereNamespace).Get(context.TODO(), KubeSphereServiceAccount, metav1.GetOptions{})
+		if err == nil {
+			secret, err := client.CoreV1().Secrets(KubesphereNamespace).Get(context.TODO(), sa.Secrets[0].Name, metav1.GetOptions{})
+			if err == nil {
+				return &Handler{
+					opt: o,
+					client: kiali.NewDefaultClient(
+						cache,
+						string(secret.Data["token"]),
+						o.KialiQueryHost,
+					),
+				}
+			}
+			klog.Warningf("get ServiceAccount's Secret failed %v", err)
+		}
+		klog.Warningf("get ServiceAccount failed %v", err)
+	}
+	// Handler should return Status code 400, instead of crash ks-apiserver
+	// when no client is defined.
+	return &Handler{opt: o, client: nil}
```

Comment:
If the client is nil, does the Handler handle the requests and return 400?

Reply:
@yuswift  The nil condition is handled bellow:
https://github.com/kubesphere/kubesphere/blob/85fd94b92245496a3e8c89cd6899e8e3355dd101/pkg/kapis/servicemesh/metrics/v1alpha2/handler.go#L160